modules/policy/engine/src/main/java/org/eclipse/sw360/antenna/policy/engine/CompareArtifactExecutor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) Bosch Software Innovations GmbH 2019.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.sw360.antenna.policy.engine;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation of the {@link RuleExecutor} for rules of type {@link CompareArtifactRule}.
+ */
+class CompareArtifactExecutor implements RuleExecutor {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SingleArtifactExecutor.class);
+
+    private final Collection<CompareArtifactRule> rules;
+
+    public CompareArtifactExecutor(final Collection<CompareArtifactRule> rules) { this.rules = rules; }
+
+    @Override
+    public Collection<ViolationIssue> executeRules(final Collection<ThirdPartyArtifact> thirdPartyArtifacts) {
+        return rules.parallelStream()
+                .map(rule -> executeRuleOnArtifacts(rule, thirdPartyArtifacts))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+    private Collection<ViolationIssue> executeRuleOnArtifacts
+            (final CompareArtifactRule rule, final Collection<ThirdPartyArtifact> thirdPartyArtifacts) {
+        return thirdPartyArtifacts.stream()
+                .map(artifact -> findViolationsForLeftHandSide(rule, artifact, thirdPartyArtifacts))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+    private Collection<ViolationIssue> findViolationsForLeftHandSide
+            (final CompareArtifactRule rule, final ThirdPartyArtifact leftHandSide, final Collection<ThirdPartyArtifact> rightHandSides) {
+
+        return rightHandSides.stream()
+                .filter(rightHandSide -> hasArtifactPolicyViolation(rule, leftHandSide, rightHandSide))
+                .map(rightHandSide -> new ViolationIssue(rule, Arrays.asList(leftHandSide, rightHandSide)))
+                .collect(Collectors.toList());
+    }
+
+    private boolean hasArtifactPolicyViolation
+            (final CompareArtifactRule rule, final ThirdPartyArtifact leftHandSide, final ThirdPartyArtifact rightHandSide) {
+        try {
+            if (!leftHandSide.equals(rightHandSide)) {
+                rule.evaluate(leftHandSide, rightHandSide);
```

Comment:
why is this realized with an Exception instead just a return value? This function just "casts" an exception back to `boolean`.